1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+will know to keep prompting for input.
+
+#### Left-associativity
+
+Like in POSIX shells, pipeline chain operators will be left associative,
+meaning chain operators will group from left to right.
+
+For example, given the following pipeline chain:
+
+```powershell
+cmd1 && cmd2 || cmd3
+```
+
+This will be grouped as:
+
+```none
+[[cmd1 && cmd2] || cmd3]
+```
+
+As a syntax tree, this would look like:
+
+```none
+            "||"
+          /      \
+        "&&"     "cmd3"
+      /      \
+   "cmd1"    "cmd2"
+```
+
+With the syntax tree deepening on the left as more operators are chained.
+
+Semantically, this means that `cmd1 && cmd2` would be evaluated first,
+and its result used to govern the evaluation of `|| cmd3`.
+
+#### Higher precedence than `&` and `;`
+
+Pipeline chain operators will have higher precedence than
+pipeline background operators (`&`)
+or statement separators (`;`).
+
+For example:
+
+```none
+cmd1 && cmd2 &
+```
+
+Will bind as:
+
+```none
+[[cmd1 && cmd2] &]
+```
+
+Having the syntax tree:
+
+```none
+            <statement>
+           /           \
+         "&&"          "&"
+        /    \
+     "cmd1"  "cmd2"
+```
+
+The consequence of this will be that an entire pipeline chain
+can be sent to a background job for evaluation,
+rather than individual pipelines within it.
+
+### Semantics
+
+#### Pipeline "success"
+
+The `&&` and `||` operators proceed based on the "success"
+of the previous pipeline.
+
+The marker of command success proposed is the `$?` automatic variable.
+This is proposed because:
+
+- It builds off an existing PowerShell concept
+- It applies to both native commands and cmdlets/functions
+- Diverging from this would create inconsistency
+- Changes to the behaviour of `$?` would be reasonably expected to change here
+
+That is, there are the following direct equivalents:
+
+```none
+cmd1 && cmd2
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 })
+```
+
+```none
+cmd1 || cmd2
+
+    |
+    v
+
+$(cmd1; if (-not $?) { cmd2 })
+```
+
+```none
+cmd1 && cmd2 || cmd3
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 }; if (-not $?) { cmd3 })
+# Note that cmd1 failing runs cmd3
+```
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+#### Pipeline output
+
+The output of a pipeline chain is the concatenation of all its pipelines,
+in the sequence of their output:
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Output 'd'
+$x # 'a','b','c','d'
+```
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Error 'Bad' || Write-Output 'd'
+# Writes the error record 'Bad'
+$x # 'a','b','c','d'
+```
+
+Commands that fail will have any pipeline output
+emitted before evaluation of the chain operator.
+
+For example:
+
+```powershell
+$x = cmd1 && cmd2
+```
+
+If `cmd1` fails but emits output, `$x` will hold that value.
+
+#### Error semantics
+
+Errors will have the same semantics as the equivalent
+`cmd1; if ($?) { cmd2 }` syntax.

```

Comment:
Expressions typically set `$?` to $true, so there will be surprising behavior:
```PowerShell
$($true || 1) -eq $($false || 1)
```

Reply:
Indeed -- perhaps I didn't address that directly enough. But my feeling is that any other logic that differs from that will confuse people and will be much more complex to maintain. My feeling is that improving the proposed experience in this respect is a matter of improving the way `$?` works.

I've also suggested expression-based `&&`/`||` equivalents (the way JavaScript does it) elsewhere, but it seems that command success is the desired feature.

In your particular example, I don't know if I'd characterise that as surprising specifically, but I think `(Invoke-Failure) || "Hello"` will be surprising, since the pipeline expression will set `$?` to true in the first instance.

Reply:
> (Invoke-Failure) || "Hello" will be surprising

Indeed!  How do we explain that behavior to folks?  Recommend that folks not use grouping/sub-expressions with this feature?  Add a PSSA rule for it?

Reply:
Yeah, I think:

- This chaining stuff depends on the `$?` concept
- Educate about `(...)` having funny effects on `$?` and by extension `&&`/`||`
- Write a PSSA rule perhaps (easy to do)
- Have a serious discussion about `$?`'s behaviour in PowerShell in an RFC (although I wouldn't be that disappointed if we keep existing behaviour -- I don't like breaking things)

I don't want to unhitch from `$?` because then we're off in the weeds with yet another thing in PowerShell we have to explain to people and why it's not consistent with the other stuff. Plus we'd have to start discussing commands where a non-zero exit code doesn't mean failure...

Reply:
Or, we could make it so that `(…)` behaves the same way as `…` when it comes to setting `$?`.

Reply:
Yeah -- that's what I mean in point 4. I think it's bigger than this syntactic feature and that change (which would be large and breaking) should be discussed separately

Reply:
It is a bit misleading that simply adding parentheses changes whether PS considers a command successful or not... We should probably revisit that. 🙂

Reply:
@vexx32: Agreed - please see #3359